lib/upipe-dveo/upipe_dveo_asi_sink.c

```diff
@@ -0,0 +1,765 @@
+    for (;;) {
+        int iovec_count = uref_block_iovec_count(uref, 0, -1);
+        if (unlikely(iovec_count == -1)) {
+            upipe_warn(upipe, "cannot read ubuf buffer");
+            break;
+        }
+        if (unlikely(iovec_count == 0)) {
+            break;
+        }
+
+        struct iovec iovecs[iovec_count];
+        if (unlikely(!ubase_check(uref_block_iovec_read(uref, 0, -1,
+                                                        iovecs)))) {
+            upipe_warn(upipe, "cannot read ubuf buffer");
+            break;
+        }
+
+        ssize_t ret = writev(upipe_dveo_asi_sink->fd, iovecs, iovec_count);
+        uref_block_iovec_unmap(uref, 0, -1, iovecs);
+
+        if (unlikely(ret == -1)) {
+            switch (errno) {
+                case EINTR:
+                    continue;
+                case EAGAIN:
+#if EAGAIN != EWOULDBLOCK
+                case EWOULDBLOCK:
+#endif
+                    //upipe_notice_va(upipe, "polling");
+                    upipe_dveo_asi_sink_poll(upipe);
+                    return false;
+                default:
+                    break;
+            }
+            upipe_warn_va(upipe, "write error to device %d (%m)", upipe_dveo_asi_sink->card_idx);
+            upipe_dveo_asi_sink_set_upump(upipe, NULL);
+            upipe_throw_sink_end(upipe);
+            break;
+        }
+
+        size_t uref_size;
+        if (ubase_check(uref_block_size(uref, &uref_size)) &&
+            uref_size == ret) {
+            /* wrote succeeded */
+            *reset_first_timestamp = false;
+            break;
+        }
+        uref_block_resize(uref, ret, -1);
+    }
+
+    return true;
+}
+
+static bool upipe_dveo_asi_sink_add_header(struct upipe *upipe, struct uref *uref,
+    uint64_t pts)
+{
+    struct upipe_dveo_asi_sink *upipe_dveo_asi_sink = upipe_dveo_asi_sink_from_upipe(upipe);
+
+    uint64_t hdr_size;
+    if (!ubase_check(uref_block_get_header_size(uref, &hdr_size)))
+        hdr_size = 0;
+
+    if (hdr_size != 0)
+        return false;
+
+    /* alloc header */
+    struct ubuf *header = ubuf_block_alloc(uref->ubuf->mgr, 8);
+    if (unlikely(!header)) {
+        return true;
+    }
+
+    /* 63-bits timestamp */
+    union {
+        uint8_t timestamp[8];
+        uint64_t pts;
+    } timestamp;
+
+    timestamp.pts = pts;
+    if (upipe_dveo_asi_sink->first_timestamp) {
+        upipe_dveo_asi_sink->first_timestamp = false;
+        // FIXME: set the counter in an empty packet, and account for latency
+        timestamp.pts |= 1LLU << 63; /* Set MSB = Set the counter */
+    }
+
+    /* write header */
+    int size = 8;
+    uint8_t *header_write_ptr;
+    if (!ubase_check(ubuf_block_write(header, 0, &size, &header_write_ptr))) {
+        upipe_err(upipe, "could not write header");
+        ubuf_free(header);
+        return true;
+    }
+    memcpy(header_write_ptr, timestamp.timestamp, 8);
```

Comment:
Are you sure the timestamp is in the native endianness ? If so, can you add a comment saying you checked ?